build-system/tasks/helpers.js

```diff
@@ -385,6 +384,49 @@ function finishBundle(srcFilename, destDir, destFilename, options) {
   }
 }
 
+/**
+ * Creates a babel plugin for esbuild for the given caller. Optionally enables
+ * caching to speed up transforms.
+ * @param {string} callerName
+ * @param {boolean} enableCache
+ * @param {function()} postStep
+ * @return {!Object}
+ */
+function getEsbuildBabelPlugin(callerName, enableCache, postStep = () => {}) {
+  return {
+    name: 'babel',
+    async setup(build) {
+      const transformContents = async (file) => {
+        const contents = await fs.promises.readFile(file.path, 'utf-8');
+        const babelOptions = babel.loadOptions({
+          caller: {name: callerName},
+          filename: file.path,
+          sourceFileName: path.basename(file.path),
+        });
+        const result = await babel.transformAsync(contents, babelOptions);
+        return {contents: result.code};
+      };
+
+      build.onLoad({filter: /.*\.[cm]?js$/, namespace: ''}, async (file) => {
+        const {path} = file;
+        if (enableCache && cache.has(path)) {
```

Comment:
are we no longer invalidating files within the cache?

Reply:
@jridgewell discovered a problem due to `esbuild` issuing multiple loads in parallel. (Lines 417-419.) Wanna handle proper invalidation via #32744 after this lands? (AFAIK, it only affects incremental building.)